### DISCUSSION TITLE
mini-batch api coverage

### DISCUSSION THREAD

**Discussion by ColdTeapot273K** [online-ml/river#818] [General]
Hey, I've made a draft implementations of mini-batch api (like `learn_many`) for several `river` abstractions like `column selector`, `transformer union` and iterating on one for `one hot encoder`. Together with already covered abstractions (`linear models` for regression/classification, `standard scaler` and `pipeline`) it should be an essential set of mini-batch tools ready. I can show here/submit a PR later if it's interesting

Mini-batch api has given me logarithmic speedup in training, from 30+ hours to 3 minutes, on some quite large dataset.

**Comment by MaxHalford** [comment]
Hey there @ColdTeapot273K! That's music to my ears. As you say, there are some essential pieces of logic that indeed should support mini-batches. Feel welcome to open a PR, we can iron out the details together.

@raphaelsty did quite a bit of work in this direction too. At some point it might be worth considering properly benchmarking these mini-batch implementations to compare them with their `*_one` alternatives. For example with a tool like [pytest-benchmark](https://pytest-benchmark.readthedocs.io/en/latest/).

**Reply by ColdTeapot273K** [reply]
@MaxHalford hey Max, a quick question: 
From what I see in the [source](https://github.com/online-ml/river/blob/main/river/linear_model/glm.py#L101-L130), the gradient updates related to `learn_many` methods function correctly if some columns are removed/reordered arbitrarily, correct? I know that `learn_one` is robust to this.

Examples:

Fitting on all these three samples in any order should produce the same coefficients, if my assumption is true:

| prefix_h | prefix_a |
| -------- | -------- |
| 0        | 0        |
| 0        | 1        |
| 0        | 0        |
| 1        | 0        |


| prefix_ff | prefix_h | prefix_a |
| --------- | -------- | -------- |
| 0         | 0        | 0        |
| 0         | 0        | 1        |
| 0         | 0        | 0        |
| 0         | 1        | 0        |


| prefix_h | prefix_a | prefix_ff |
| -------- | -------- | --------- |
| 0        | 0        | 0         |
| 0        | 1        | 0         |
| 0        | 0        | 0         |
| 1        | 0        | 0         |


---
I'm trying to make sure that my implementation of sparse modality for mini-batch one-hot encoder does not violate any guarantees expected by the `glm.py`

**Reply by MaxHalford** [reply]
You're correct @ColdTeapot273K, processing these 3 dataframes in any order should produce the same model state. You could even add a unit test to make sure of this.

**Reply by ColdTeapot273K** [reply]
@MaxHalford noice 👌

Funnily enough, when writing that message i've figured it's trivial to test, so implemented a test and verified

**Reply by ColdTeapot273K** [reply]
Finally:

#827
